### DISCUSSION TITLE
How to reset data from database in order to get defaultValues working

### DISCUSSION THREAD

**Discussion by Pietermachiel** [react-hook-form/react-hook-form#2622] [Q&A]
In a MERN project I want to update .../api/recipes/:id but 'defaultValues' is not working. I know that if `defaultValues` is not read on the initial load, you will need to use `reset({ defaultValues })` at useEffect, but I am not sure how to do that. 

My code is not filling the form, but when manually changing all the values, the update is succesfull.

Can somebody give me a code example, that fills the form initially with the data from database?

This is my code:

 const [therecipe, setTheRecipe] = useState([]);

 useEffect(() => {
    async function getData() {
      const res = await fetch(`${apiUrl}/recipes/${API}`);
      res.json().then((res) => setTheRecipe(res));
    }
    getData();
  }, [API]);

  const { register, control, handleSubmit, reset, errors } = useForm({
    defaultValues: {
      _id: therecipe._id,
      title: therecipe.title,
      dish: therecipe.dish,
      tags: therecipe.tags,
    },
  });

  useEffect(() => {
    reset( therecipe );
  }, [reset]);

  console.log("therecipe");
  console.log(therecipe);

**Comment by bluebill1049** [comment]
```
useEffect(() => {
  async function getData() {
     const res = await fetch(${apiUrl}/recipes/${API});
     res.json().then((res) => reset( therecipe ););
  }
  getData();
}, [API]);
```

**Reply by Pietermachiel** [reply]
Hello Bluebill, thanks for your quick reply. I put your code directly after the useEffect() fetch call to the server and it causes a lot of rerenders, but it does not effect the defaultValues filling the form initially (in order to edit them by the user).

Did I misunderstood?

I did it like this:
```
  useEffect(() => {
    async function getData() {
      const res = await fetch(`${apiUrl}/recipes/${API}`);
      res.json().then((res) => setTheRecipe(res));
    }
    getData();
  }, [API]);

  useEffect(() => {
    async function getData() {
      const res = await fetch(`${apiUrl}/recipes/${API}`);
      res.json().then((res) => reset(therecipe));
    }
    getData();
  }, [API]);
```

**Reply by bluebill1049** [reply]
just do one without the request and see if that works for you. i don't understand many render part as well.

```
useEffect(() => {
  setTimeout(() => {
     reset({
       values,
     })
  }, 5000)
}, [API]);
```

**Comment by Pietermachiel** [comment]
Well, the only difference now is the delay of 5000. Why can't 'defaultValues' sort of wait until the request, or the reset is done?

**Reply by bluebill1049** [reply]
what do you mean by why can't it wait? the setTimeout was basically mimicking a request send out and then get resolved.

**Comment by Pietermachiel** [comment]
Well, with and without the 'reset' (reset both with and without the request), the request of 'therecipe' is being resolved, as console.log(therecipe) shows.

So my question is: why does defaultValues:{therecipe} does not do the job?

Here's my code with your remark:

const [therecipe, setTheRecipe] = useState([]);

  const API = props.location.state;

  useEffect(() => {
    async function getData() {
      const res = await fetch(`${apiUrl}/recipes/${API}`);
      res.json().then((res) => setTheRecipe(res));
    }
    getData();
  }, [API]);

  useEffect(() => {
    setTimeout(() => {
      reset({
        therecipe,
      });
    }, 5000);
  }, [API]);

  const { register, control, handleSubmit, reset, errors } = useForm({
    defaultValues: {
      _id: therecipe._id,
      title: therecipe.title,
      dish: therecipe.dish,
      tags: therecipe.tags,
      related: therecipe.related,
      fresh: therecipe.fresh,
      stock: therecipe.stock,
      directions: therecipe.directions,
      book: therecipe.book,
      info: therecipe.info,
      date: therecipe.date,
    },
  });

  console.log(therecipe);

console.log(therecipe
{tags: Array(1), _id: "5f39029210df671b19ad01a5", title: "Canneloni", dish: {…}, book: {…}, …}
book: {_id: "5f3403e4b7cea1a9743316c1", name: "Italian food", kitchen: "5f1e8da2498d92a27abdc21c", author: "Elizabeth David", publisher: "MacDonald", …}
date: []
directions: [{…}]
dish: {_id: "5f3406f7b7cea1a9743316c5", name: "pasta", __v: 0}
fresh: [{…}]
info: "Canneloncannelon"
related: [{…}]
stock: [{…}]
tags: [{…}]
title: "Canneloncannelonzzz"
_id: "5f39029210df671b19ad01a5"
__proto__: Object

**Reply by bluebill1049** [reply]
because we cache `defaultValues`, it's the default values for the form, if you want to update the form async then use `reset(values)`

**Reply by Pietermachiel** [reply]
I am afraid we talk past each other, sorry about that.

I was confused by earlier posts on this matter, but my 'matter' seems to be different :-)

What I seem to understand now, is that reset would blank the form, right? But that is not what i need.

In my app when the user hits the link in recipes to a recipe that he wants to edit, the form is not filled in the first place. I maybe false conclusion was that this is because defaultValues: {...} does not fill the form.

So why would the form not be filled by the api call?

Thanks, and sorry again about my misunderstanding

**Reply by bluebill1049** [reply]
`reset` would not blank the form, take a look at the doc: https://react-hook-form.com/api#reset I am keeping mentioning above, you can supply an argument and it will set up all the form values.

if you want to update individual input then use `setValue`

**Reply by Pietermachiel** [reply]
Ok, I will keep on trying. Thanks for your patience.

**Reply by Pietermachiel** [reply]
Hi Bluebill,

I cannot get it working with your ‘reset solution’, but I found a workaround by filtering ‘therecipe’ from the API endpoint of ‘recipes’ and pass it as a prop from a lifted component. This works! Not ideal because I have to load the recipes with all the properties now, which makes it expensive so I am still hoping of a solution. I just find it counter intuitive that useForm(defaultValues) caches before React [therecipe, setTheRecipe] = setState([]) is set by useEffect( ... fetch(...) ... ). In my situation and by all means defaultValues is not taking ‘therecipe’ properties. Hope I made myself clear (sorry I am not a real programmer :-) Can you have another look?

**Reply by bluebill1049** [reply]
please provide a codesandbox.

**Reply by Pietermachiel** [reply]
Here is my codesandbox: https://codesandbox.io/s/kind-mountain-plcv1
EditRecipe.js component contains my solution, the ‘useForm reset' solution (as far as I have understood :-) is commented out.

**Reply by Pietermachiel** [reply]
Hello bluebill, can you please have another look, or is my codesandbox not working for you?

**Comment by Pietermachiel** [comment]
Hello, 

I provided a codesandbox a few days ago.

Codesandbox: https://codesandbox.io/s/kind-mountain-plcv1 <https://codesandbox.io/s/kind-mountain-plcv1>

EditRecipe.js component contains my workaround solution.
The ‘useForm reset' solution (as far as I have understood :-), which is not working for me, is commented out.

Could you please have another look?

Best whishes, PIeter Roozen

_____
 
Pieter Roozen
 
Carel Reinierszkade 25
2593 HK Den Haag
The Netherlands
 
+31 (0)6 81883838
pieter@roozen.nl
 
www.roozen.nl

Is printen, echt nodig?

**Comment by Pietermachiel** [comment]
Why no response? Is something wrong?
